### PR TITLE
[SPARK-30582][WEBUI] Spark UI is not showing Aggregated Metrics by Executor in stage page

### PR DESCRIPTION
**Pull request by SaurabhChawla100**
### What changes were proposed in this pull request?

There are scenarios where Spark History Server is located behind the VPC. So whenever api calls hit to get the executor Summary(allexecutors). There can be delay in getting the response of executor summary and in mean time "stage-page-template.html" is loaded and the response of executor Summary is not added to the stage-page-template.html.

As the result of which Aggregated Metrics by Executor in stage page is showing blank.

This scenario can be easily found in the cases when there is some proxy-server which is responsible for sending the request and response to spark History server.
This can be reproduced in Knox/In-house proxy servers which are used to send and receive response to Spark History Server.

Alternative scenario to test this case, Open the spark UI in developer mode in browser add some breakpoint in stagepage.js, this will add some delay in getting the response and now if we check the spark UI for stage Aggregated Metrics by Executor in stage page is showing blank.

So In-order to fix this there is a need to add the change in stagepage.js . There is a need to add the api call to get the html page(stage-page-template.html) first and after that other api calls to get the data that needs to attached in the stagepage (like executor Summary, stageExecutorSummaryInfoKeys exc)

### Why are the changes needed?
Since stage page is useful for debugging purpose, This helps in understanding how many task ran on the particular executor and information related to shuffle read and write on that executor. 

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Manually tested. Testing this in a reproducible way requires a running browser or HTML rendering engine that executes the JavaScript.Open the spark UI in developer mode in browser add some breakpoint in stagepage.js, this will add some delay in getting the response and now if we check the spark UI for stage Aggregated Metrics by Executor in stage page is showing blank.

Before fix 

Screenshot 2020-01-20 at 3 21 55 PM

After fix 

Screenshot 2020-01-20 at 3 23 12 PM

### CODE REVIEW

**Comment by SaurabhChawla100** [comment]
cc @srowen @dongjoon-hyun

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js [hunk 1]
```diff
@@ -473,27 +492,6 @@ $(document).ready(function () {
             var accumulatorTable = responseBody.accumulatorUpdates.filter(accumUpdate =>
                 !(accumUpdate.name).toString().includes("internal."));
 
-            // rendering the UI page
-            var data = {"executors": response};
```
Comment (srowen) [inline core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: Maybe a dumb question, but where did this go then - how is the `data` loaded and displayed?

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js [hunk 2]
```diff
@@ -325,6 +325,25 @@ $(document).ready(function () {
...
     getStandAloneAppId(function (appId) {
+        // rendering the UI page
+        $.get(createTemplateURI(appId, "stagespage"), function(template) {
+          tasksSummary.append(Mustache.render($(template).filter("#stages-summary-template").html()));
+
+          $("#additionalMetrics").click(function(){
+              $("#arrowtoggle1").toggleClass("arrow-open arrow-closed");
+              $("#toggle-metrics").toggle();
+              if (window.localStorage) {
+                        window.localStorage.setItem("arrowtoggle1class", $("#arrowtoggle1").attr('class'));
```
Comment (tgravescs) [inline core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: fix spacing

**Review by tgravescs — COMMENTED** [review]
So does this result in the page showing up as empty tables or blank first and then when the queries come back they get filled in?

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js (hunk 1 above)
Comment (SaurabhChawla100) [reply core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: Before Fix- In existing code

`var data = {"executors": response};
            $.get(createTemplateURI(appId, "stagespage"), function(template) {
                tasksSummary.append(Mustache.render($(template).filter("#stages-summary-template").html(), data));`

In code response object comes from this api call ("$.getJSON(endPoint + "/" + stageAttemptId, function(response, status, jqXHR)") and this API call is invoked before the call made to get the html page of stage and then adding that response in the html page.

After Fix 

The stage page html is first loaded and all the api calls are inside the `$.get(createTemplateURI(appId, "stagespage"), function(template) {` , So we dont need to add the response here in this scenario.

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js (hunk 2 above)
Comment (SaurabhChawla100) [reply core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: fixed spacing

**Comment by SaurabhChawla100** [comment]
> So does this result in the page showing up as empty tables or blank first and then when the queries come back they get filled in?

@tgravescs  - Yes the page first showed up with empty tables and once the data comes these will be filled

Please find the snapshot of all the events from debugger in developer mode 

1) Stage Page with empty tables

Screenshot 2020-01-24 at 7 59 46 AM

2) Stage page with some data after the one of the api calls

Screenshot 2020-01-24 at 8 00 18 AM

3)Final Stage page 

Screenshot 2020-01-24 at 8 01 02 AM

**Comment by tgravescs** [comment]
ok to test

**Comment by tgravescs** [comment]
cc @abellina

**Comment by tgravescs** [comment]
test this please

**Review by tgravescs — APPROVED** [review]

**Comment by SaurabhChawla100** [comment]
@srowen - Please review the code change as you are also in the reviewer list , Already got one approval from @tgravescs .Not sure about the process to get the PR merged, Since its my first PR here. Thank-you

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**